Synchrona rev b leds service

tachicialex opened the pull request:
c5f0b8f9c05b - synchrona: move LEDs control to separate service
f1ffa358a033 - synchrona: add direction to script

Tested with the latest Kuiper Image + newest rpi-ad9545-hmc7044 /kernel image/modules.

tachicialex committed:
installers: ad-synchrona-14-leds: add direction to script

Write the GPIO direction in the loop. Otherwise sometimes
we will get 'Operation not permited'.

Fixes: 2a290d6f2086 ("synchrona: add Status LEDs control")
Signed-off-by: Alexandru Tachici <alexandru.tachici@analog.com>

tachicialex committed:
installers: common: move LEDs control to separate service

Starting the looping script with the ExecStartPost
will make the main service to restart every 90 sec. due
to timeout.

Fixes: 2a290d6f2086 ("synchrona: add Status LEDs control")
Signed-off-by: Alexandru Tachici <alexandru.tachici@analog.com>

nunojsa commented on installers/ad-synchrona-14-leds.sh:
```diff
@@ -55,6 +55,9 @@ do
 	fi
 
+	echo "out" > "/sys/class/gpio/gpio$RED_LED1/direction"
+	echo "out" > "/sys/class/gpio/gpio$GREEN_LED1/direction"
```
Yeah, I saw a similar thing on the main service. While in there it's more odd to me what could have been the issue/race, in here I can see what could be the race... When we reload (remove + aload) the overlay and the ad9545 device is unbind, the pinctrl settings are also removed which brings the pins back to their default (which might be input). If you happen to try to set a value at that moment, the issue pops up...

So yeah, for now let's have this workaround and keep things simple but I'm not sure if we would need some kind of synchronization between this script and the reload_dtb.sh

nunojsa commented on installers/common.sh:
```diff
@@ -204,20 +206,37 @@ function _setup_synchrona_service() {
+    ExecStartPre=-/usr/bin/sh -c "/usr/bin/echo 16 > /sys/class/gpio/export"
+    ExecStart=/usr/bin/bash /etc/raspap/synchrona/ad-synchrona-14-leds.sh > /dev/null 2>&1 &
     ExecStopPost=-/usr/bin/sh -c "/usr/bin/echo 12 > /sys/class/gpio/unexport
     ExecStopPost=-/usr/bin/sh -c "/usr/bin/echo 16 > /sys/class/gpio/unexport
```
This is missing `"` in the end. In fact, this is a bug introduced when the led script was added (which I completely failed to see during review). So, the fix for this should be in it's own patch before doing moving this to it's own service...

nunojsa commented on installers/common.sh:
```diff
@@ -204,20 +206,37 @@ function _setup_synchrona_service() {
+    Type=simple
+    User=root
+    ExecStartPre=-/usr/bin/sh -c "/usr/bin/echo 12 > /sys/class/gpio/export"
+    ExecStartPre=-/usr/bin/sh -c "/usr/bin/echo 16 > /sys/class/gpio/export"
```
nit: Can't we do this in the script? So we check if the gpio is already exported and if not, we export them. The same goes for the unexport part that we can do when leaving the script (you might need to trap SIGSTOP in order to terminate the main loop). In this way we could leave the service file a bit more simple...

nunojsa commented on installers/common.sh:
```diff
@@ -204,20 +206,37 @@ function _setup_synchrona_service() {
+    User=root
+    ExecStartPre=-/usr/bin/sh -c "/usr/bin/echo 12 > /sys/class/gpio/export"
+    ExecStartPre=-/usr/bin/sh -c "/usr/bin/echo 16 > /sys/class/gpio/export"
+    ExecStart=/usr/bin/bash /etc/raspap/synchrona/ad-synchrona-14-leds.sh > /dev/null 2>&1 &
```
Do we still need to run the script in the background? Im fairly sure that systemd will already "daemonize" the script... I would also not redirect all the log to /dev/null as it might be useful to run `journalctl -u synchrona_leds` to look for possible issues

nunojsa reviewed (commented):
Just some remarks from me...

tachicialex committed:
installers: common: fix missing quotes

Unexport commands for GPIOs 12 and 16 had missing quotes.

Fixes: 2a290d6f2086 ("synchrona: add Status LEDs control")
Signed-off-by: Alexandru Tachici <alexandru.tachici@analog.com>

tachicialex committed:
installers: common: Removed pre/post from LEDs service

Removed ExecStartPre/Post commands from the synchrona_leds.service.
Added same behaviour to the script ran by the service.

Signed-off-by: Alexandru Tachici <alexandru.tachici@analog.com>

tachicialex commented:
e2a4bad9465d - installers: common: Removed pre/post from LEDs service
cb3e656155df - installers: common: move LEDs control to separate service
71d198cedb1c - installers: common: fix missing quotes
d962b41c123d - installers: ad-synchrona-14-leds: add direction to script

Changelog v1 - > v2:
- fixed missing " at the end of unexport
- ad-synchrona-14-leds.sh don't run script in background
- no longer redirect ad-synchrona-14-leds.sh output to /dev/null, for visibility in logs
- moved export/unexport from service file to ad-synchrona-14-leds.sh

Tested with latest kuiper + dt

nunojsa commented on installers/ad-synchrona-14-leds.sh:
```diff
@@ -17,6 +17,24 @@ AD9545_LOCKED=0
 AD9545_FREERUN=0
 
+trap synchrona_leds_exit SIGTERM
+function synchrona_leds_exit()
```
nit: typically in bash we don't really use `function`. Just `synchrona_leds_exit() {`

nunojsa commented on installers/ad-synchrona-14-leds.sh:
```diff
@@ -17,6 +17,24 @@ AD9545_LOCKED=0
+{
+	echo $RED_LED1 > "/sys/class/gpio/unexport"
+	echo $GREEN_LED1 > "/sys/class/gpio/unexport"
+	exit
```
I was more thinking in having a control variable to break the loop. But yeah, likely is not a big deal but feels wrong to exit a program in a signal handler (even though it's a script)

nunojsa commented on installers/ad-synchrona-14-leds.sh:
```diff
@@ -17,6 +17,24 @@ AD9545_LOCKED=0
+if [ ! -d "/sys/class/gpio/gpio$RED_LED1" ]
+then
+	echo $RED_LED1 > "/sys/class/gpio/export"
+fi
```
nit: this could be a one liner `[ ! -d "/sys/class/gpio/gpio$RED_LED1" ] && echo $RED_LED1 > "/sys/class/gpio/export"`

nunojsa reviewed (approved):
Just some minor nits that you can ignore if you want... Anyways, you need to rebase and surely test it :)

tachicialex commented:
Changelog v2 -> v3:
- rebased over synchrona_revB
- removed "function"
- moved all "if -d" gpio dir checking to oneliners
- added a control var. to exit while() loop nicer

Pushed here just for testing. Not tested yet.

nunojsa reviewed (approved):
Let me know when it's tested...

tachicialex commented:
Ok, script is working well, when service gets stopped, LED turns red and script process gets killed. Next start ,script process is back running and changing the LED color accordingly. Tested today.
